internal/driver/custommetadata.go

```diff
@@ -0,0 +1,84 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+//
+// Copyright (C) 2022 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package driver
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/edgexfoundry/go-mod-core-contracts/v2/errors"
+	contract "github.com/edgexfoundry/go-mod-core-contracts/v2/models"
+)
+
+// setCustomMetadata will return a map containing the fields provided in the call to the function
+func (onvifClient *OnvifClient) setCustomMetadata(device contract.Device, data []byte) (contract.Device, errors.EdgeX) {
+	var dataObj contract.ProtocolProperties
+
+	err := json.Unmarshal(data, &dataObj)
+	if err != nil {
+		return device, errors.NewCommonEdgeX(errors.KindServerError, "failed to unmarshal the json request body", err)
+	}
+	if len(dataObj) == 0 {
+		return device, errors.NewCommonEdgeX(errors.KindServerError, "no data in PUT command", err)
+	}
+	for key, value := range dataObj {
+		trimValue := strings.TrimSpace(value)
+		trimKey := strings.TrimSpace(key)
+		if trimValue == "delete" || trimValue == "Delete" {
+			delete(device.Protocols[CustomMetadata], trimKey)
+			continue
+		}
+		if len(trimKey) == 0 {
+			continue
+		}
+		device.Protocols[CustomMetadata][trimKey] = trimValue // create or update a field in CustomMetadata
+	}
+
+	return device, nil
+}
+
+func (onvifClient *OnvifClient) getCustomMetadata(device contract.Device, data []byte) (contract.ProtocolProperties, errors.EdgeX) {
+	var metadataObj contract.ProtocolProperties
+	var err error
+
+	if len(data) == 0 { // if no list is provided, return all
+		metadataObj = device.Protocols[CustomMetadata]
+	} else { // if a list of fields is provided, return those specific fields
+		metadataObj, err = onvifClient.getSpecificCustomMetadata(device, data)
+		if err != nil {
+			return nil, errors.NewCommonEdgeX(errors.KindServerError, fmt.Sprintf("failed to get specific metadata for device %s", onvifClient.DeviceName), err)
+		}
+	}
+
+	return metadataObj, nil
+}
```

Comment:
I think this can be simplified

```suggestion
func (onvifClient *OnvifClient) getCustomMetadata(device contract.Device, data []byte) (contract.ProtocolProperties, errors.EdgeX) {
	if len(data) == 0 { // if no list is provided, return all
		return device.Protocols[CustomMetadata], nil
	}

	// if a list of fields is provided, return those specific fields
	metadataObj, err := onvifClient.getSpecificCustomMetadata(device, data)
	if err != nil {
		return nil, errors.NewCommonEdgeX(errors.KindServerError, fmt.Sprintf("failed to get specific metadata for device %s", onvifClient.DeviceName), err)
	}
	return metadataObj, nil
}
```